Sources/Authentication/Credentials/Token.swift

```diff
@@ -0,0 +1,55 @@
+// MARK: Data structure
+
+public struct Token: Crendentials {
+    public let string: String
+
+    public init(string: String) {
+        self.string = string
+    }
+}
+
+// MARK: Authenticatable
+
+public protocol TokenAuthenticatable: Authenticatable {
+    /// The token entity that contains a foreign key
+    /// pointer to the user table
+    associatedtype TokenType: TokenProtocol
+
+    /// Returns the user matching the supplied token.
+    static func authenticate(_: Token) throws -> Self
+}
+
+public protocol TokenProtocol {
+    static var tokenKey: String { get }
+    static func findUser<U: Entity>(for: Token) throws -> U
+}
+
+extension TokenProtocol {
+    public static var tokenKey: String {
+        return "token"
+    }
+}
+
+// MARK: Entity conformance
+
+import Fluent
+
+extension TokenAuthenticatable where Self: Entity {
+    public static func authenticate(_ token: Token) throws -> Self {
+        return try TokenType.findUser(for: token)
+    }
+}
+
+extension TokenProtocol where Self: Entity {
```

Comment:
I think this sort of supports in my argument for making one path of auth, I think right now TokenProtocol(TokenAuthenticator) and TokenAuthenticatable don't have a clear separation of responsibilities. I think we can focus it.